File: data_ruby/requester.rb
require 'base64'
require 'json'
require 'dropbox-sign'

class Requester
  FILE_UPLOADS_DIR = __dir__ + '/../file_uploads'

  attr_accessor :auth_type

  attr_accessor :auth_key

  attr_accessor :data

  attr_accessor :dev_mode

  attr_accessor :files

  attr_accessor :header_params

  attr_accessor :operation_id

  attr_accessor :parameters

  attr_accessor :api_server

  def initialize(
    auth_type = '',
    auth_key = '',
    api_server = '',
    json_data = '',
    dev_mode = nil
  )
    self.auth_type = auth_type.downcase
    self.auth_key = auth_key
    self.api_server = api_server
    self.dev_mode = self.to_boolean(dev_mode)

    self.read_json_data(json_data)
  end

  def run
    begin
      body, status_code, headers = call_from_operation_id

      response = {
        body: body.to_body,
        status_code: status_code,
        headers: get_response_headers(headers),
      }

      puts JSON.pretty_generate(response)
    rescue Dropbox::Sign::ApiError => e
      response = {
        body: e.response_body.to_body,
        status_code: e.code,
        headers: get_response_headers(e.response_headers),
      }

      puts JSON.pretty_generate(response)
    end
  end

  private

  def get_client
    config = Dropbox::Sign::Configuration.new
    config.host = self.api_server

    if self.auth_type === 'apikey'
      config.username = self.auth_key
    elsif self.auth_type === 'oauth'
      config.access_token = self.auth_key
    else
      raise 'Invalid auth type. Must be "apikey" or "oauth".'
    end

    self.header_params = {}

    if self.dev_mode
      self.header_params = {"Cookie" => 'XDEBUG_SESSION=xdebug'}
    end

    Dropbox::Sign::ApiClient.new(config)
  end

  def read_json_data(base64_json)
    if base64_json.is_a?(String) && !base64_json.to_s.strip.empty?
      begin
        json = JSON.parse(Base64.decode64(base64_json), :symbolize_names => true)
      rescue StandardError
        raise 'Invalid base64 JSON data provided.'
      end
    else
      raise 'No valid JSON data provided.'
    end

    self.operation_id = json[:operationId]
    self.data = json[:data] || {}
    self.files = JSON.parse(json[:files].to_json, :symbolize_names => false) || {}
    self.parameters = JSON.parse(json[:parameters].to_json, :symbolize_names => false) || {}
  end

  def call_from_operation_id
    response = self.account_api
    if response
      return response
    end

    response = self.api_app_api
    if response
      return response
    end

    response = self.bulk_send_job_api
    if response
      return response
    end

    response = self.embedded_api
    if response
      return response
    end

    response = self.oauth_api
    if response
      return response
    end

    response = self.report_api
    if response
      return response
    end

    response = self.signature_request_api
    if response
      return response
    end

    response = self.team_api
    if response
      return response
    end

    response = self.template_api
    if response
      return response
    end

    response = self.unclaimed_draft_api
    if response
      return response
    end

    raise 'Invalid operation_id ' + self.operation_id
  end

  def get_response_headers(headers)
    formatted = {}

    headers.each do |k, v|
      formatted[k.downcase] = v
    end

    formatted
  end

  def get_file(name)
    if name.is_a?(String) && self.files.key?(name)
      File.new(FILE_UPLOADS_DIR + '/' + self.files[name], 'r')
    end
  end

  def get_files(name)
    if name.is_a?(String) && self.files.key?(name) && self.files[name].length > 0
      files = []

      self.files[name].each do |file|
        files.append(File.new(FILE_UPLOADS_DIR + '/' + file, 'r'))
      end

      files
    end
  end

  def account_api
    api_client = self.get_client
    api = Dropbox::Sign::AccountApi.new(api_client)

    if self.operation_id === 'accountCreate'
      obj = Dropbox::Sign::AccountCreateRequest.init(self.data)

      return api.account_create_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'accountGet'
      return api.account_get_with_http_info(
        {
          header_params: self.header_params,
          account_id: self.parameters['account_id'] || nil,
          email_address: self.parameters['email_address'] || nil,
        }
      )
    end

    if self.operation_id === 'accountUpdate'
      obj = Dropbox::Sign::AccountUpdateRequest.init(self.data)

      return api.account_update_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'accountVerify'
      obj = Dropbox::Sign::AccountVerifyRequest.init(self.data)

      return api.account_verify_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end
  end

  def api_app_api
    api_client = self.get_client
    api = Dropbox::Sign::ApiAppApi.new(api_client)

    if self.operation_id === 'apiAppCreate'
      obj = Dropbox::Sign::ApiAppCreateRequest.init(self.data)
      obj.custom_logo_file = self.get_file('custom_logo_file')

      return api.api_app_create_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'apiAppDelete'
      return api.api_app_delete_with_http_info(
        self.parameters['client_id'],
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'apiAppGet'
      return api.api_app_get_with_http_info(
        self.parameters['client_id'],
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'apiAppList'
      return api.api_app_list_with_http_info(
        {
          header_params: self.header_params,
          page: self.parameters['page'] || 1,
          page_size: self.parameters['page_size'] || 20,
        }
      )
    end

    if self.operation_id === 'apiAppUpdate'
      obj = Dropbox::Sign::ApiAppUpdateRequest.init(self.data)
      obj.custom_logo_file = self.get_file('custom_logo_file')

      return api.api_app_update_with_http_info(
        self.parameters['client_id'],
        obj,
        {
          header_params: self.header_params,
        }
      )
    end
  end

  def bulk_send_job_api
    api_client = self.get_client
    api = Dropbox::Sign::BulkSendJobApi.new(api_client)

    if self.operation_id === 'bulkSendJobGet'
      return api.bulk_send_job_get_with_http_info(
        self.parameters['bulk_send_job_id'],
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'bulkSendJobList'
      return api.bulk_send_job_list_with_http_info(
        {
          header_params: self.header_params,
          page: self.parameters['page'] || 1,
          page_size: self.parameters['page_size'] || 20,
        }
      )
    end
  end

  def embedded_api
    api_client = self.get_client
    api = Dropbox::Sign::EmbeddedApi.new(api_client)

    if self.operation_id === 'embeddedEditUrl'
      obj = Dropbox::Sign::EmbeddedEditUrlRequest.init(self.data)

      return api.embedded_edit_url_with_http_info(
        self.parameters['template_id'],
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'embeddedSignUrl'
      return api.embedded_sign_url_with_http_info(
        self.parameters['signature_id'],
        {
          header_params: self.header_params,
        }
      )
    end
  end

  def oauth_api
    api_client = self.get_client
    api = Dropbox::Sign::OAuthApi.new(api_client)

    if self.operation_id === 'oauthTokenGenerate'
      obj = Dropbox::Sign::OAuthTokenGenerateRequest.init(self.data)

      return api.oauth_token_generate_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'oauthTokenRefresh'
      obj = Dropbox::Sign::OAuthTokenRefreshRequest.init(self.data)

      return api.oauth_token_refresh_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end
  end

  def report_api
    api_client = self.get_client
    api = Dropbox::Sign::ReportApi.new(api_client)

    if self.operation_id === 'reportCreate'
      obj = Dropbox::Sign::ReportCreateRequest.init(self.data)

      return api.report_create_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end
  end

  def signature_request_api
    api_client = self.get_client
    api = Dropbox::Sign::SignatureRequestApi.new(api_client)

    if self.operation_id === 'signatureRequestBulkCreateEmbeddedWithTemplate'
      obj = Dropbox::Sign::SignatureRequestBulkCreateEmbeddedWithTemplateRequest.init(self.data)

      obj.signer_file = self.get_file('signer_file')

      return api.signature_request_bulk_create_embedded_with_template_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestBulkSendWithTemplate'
      obj = Dropbox::Sign::SignatureRequestBulkSendWithTemplateRequest.init(self.data)

      obj.signer_file = self.get_file('signer_file')

      return api.signature_request_bulk_send_with_template_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestCancel'
      return api.signature_request_cancel_with_http_info(
        self.parameters['signature_request_id'],
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestCreateEmbedded'
      obj = Dropbox::Sign::SignatureRequestCreateEmbeddedRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.signature_request_create_embedded_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestCreateEmbeddedWithTemplate'
      obj = Dropbox::Sign::SignatureRequestCreateEmbeddedWithTemplateRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.signature_request_create_embedded_with_template_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestFilesAsFileUrl'
      return api.signature_request_files_as_file_url_with_http_info(
        self.parameters['signature_request_id'],
        {
          header_params: self.header_params
        }
      )
    end

    if self.operation_id === 'signatureRequestGet'
      return api.signature_request_get_with_http_info(
        self.parameters['signature_request_id'],
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestList'
      return api.signature_request_list_with_http_info(
        {
          header_params: self.header_params,
          account_id: self.parameters['account_id'] || nil,
          page: self.parameters['page'] || 1,
          page_size: self.parameters['page_size'] || 20,
          query: self.parameters['query'] || nil,
        }
      )
    end

    if self.operation_id === 'signatureRequestReleaseHold'
      return api.signature_request_release_hold_with_http_info(
        self.parameters['signature_request_id'],
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestRemind'
      obj = Dropbox::Sign::SignatureRequestRemindRequest.init(self.data)

      return api.signature_request_remind_with_http_info(
        self.parameters['signature_request_id'],
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestRemove'
      return api.signature_request_remove_with_http_info(
        self.parameters['signature_request_id'],
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestSend'
      obj = Dropbox::Sign::SignatureRequestSendRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.signature_request_send_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestSendWithTemplate'
      obj = Dropbox::Sign::SignatureRequestSendWithTemplateRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.signature_request_send_with_template_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'signatureRequestUpdate'
      obj = Dropbox::Sign::SignatureRequestUpdateRequest.init(self.data)

      return api.signature_request_update_with_http_info(
        self.parameters['signature_request_id'],
        obj,
        {
          header_params: self.header_params,
        }
      )
    end
  end

  def team_api
    api_client = self.get_client
    api = Dropbox::Sign::TeamApi.new(api_client)

    if self.operation_id === 'teamAddMember'
      obj = Dropbox::Sign::TeamAddMemberRequest.init(self.data)

      return api.team_add_member_with_http_info(
        obj,
        {
          header_params: self.header_params,
          team_id: self.parameters['team_id'],
        }
      )
    end

    if self.operation_id === 'teamCreate'
      obj = Dropbox::Sign::TeamCreateRequest.init(self.data)

      return api.team_create_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'teamDelete'
      return api.team_delete_with_http_info(
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'teamGet'
      return api.team_get_with_http_info(
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'teamRemoveMember'
      obj = Dropbox::Sign::TeamRemoveMemberRequest.init(self.data)

      return api.team_remove_member_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'teamUpdate'
      obj = Dropbox::Sign::TeamUpdateRequest.init(self.data)

      return api.team_update_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end
  end

  def template_api
    api_client = self.get_client
    api = Dropbox::Sign::TemplateApi.new(api_client)

    if self.operation_id === 'templateAddUser'
      obj = Dropbox::Sign::TemplateAddUserRequest.init(self.data)

      return api.template_add_user_with_http_info(
        self.parameters['template_id'],
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'templateCreateEmbeddedDraft'
      obj = Dropbox::Sign::TemplateCreateEmbeddedDraftRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.template_create_embedded_draft_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'templateDelete'
      return api.template_delete_with_http_info(
        self.parameters['template_id'],
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'templateFilesAsFileUrl'
      return api.template_files_as_file_url_with_http_info(
        self.parameters['template_id'],
        {
          header_params: self.header_params
        }
      )
    end

    if self.operation_id === 'templateGet'
      return api.template_get_with_http_info(
        self.parameters['template_id'],
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'templateList'
      return api.template_list_with_http_info(
        {
          header_params: self.header_params,
          account_id: self.parameters['account_id'] || nil,
          page: self.parameters['page'] || 1,
          page_size: self.parameters['page_size'] || 20,
          query: self.parameters['query'] || nil,
        }
      )
    end

    if self.operation_id === 'templateRemoveUser'
      obj = Dropbox::Sign::TemplateRemoveUserRequest.init(self.data)

      return api.template_remove_user_with_http_info(
        self.parameters['template_id'],
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'templateUpdateFiles'
      obj = Dropbox::Sign::TemplateUpdateFilesRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.template_update_files_with_http_info(
        self.parameters['template_id'],
        obj,
        {
          header_params: self.header_params,
        }
      )
    end
  end

  def unclaimed_draft_api
    api_client = self.get_client
    api = Dropbox::Sign::UnclaimedDraftApi.new(api_client)

    if self.operation_id === 'unclaimedDraftCreate'
      obj = Dropbox::Sign::UnclaimedDraftCreateRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.unclaimed_draft_create_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'unclaimedDraftCreateEmbedded'
      obj = Dropbox::Sign::UnclaimedDraftCreateEmbeddedRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.unclaimed_draft_create_embedded_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'unclaimedDraftCreateEmbeddedWithTemplate'
      obj = Dropbox::Sign::UnclaimedDraftCreateEmbeddedWithTemplateRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.unclaimed_draft_create_embedded_with_template_with_http_info(
        obj,
        {
          header_params: self.header_params,
        }
      )
    end

    if self.operation_id === 'unclaimedDraftEditAndResend'
      obj = Dropbox::Sign::UnclaimedDraftEditAndResendRequest.init(self.data)
      obj.files = self.get_files('files')

      return api.unclaimed_draft_edit_and_resend_with_http_info(
        self.parameters['signature_request_id'],
        obj,
        {
          header_params: self.header_params,
        }
      )
    end
  end

  def to_boolean(s)
    (s == true || s == 'true' || s == 1)
  end
end

requester = Requester.new(
  ENV['AUTH_TYPE'] || '',
  ENV['AUTH_KEY'] || '',
  ENV['API_SERVER'] || '',
  ENV['JSON_DATA'] || '',
  ENV['DEV_MODE'] || false,
)

requester.run

File: data_java/Requester.java
package com.dropbox.sign.sdk_tester;

import com.dropbox.sign.ApiClient;
import com.dropbox.sign.ApiException;
import com.dropbox.sign.ApiResponse;
import com.dropbox.sign.Configuration;
import com.dropbox.sign.api.*;
import com.dropbox.sign.auth.HttpBasicAuth;
import com.dropbox.sign.auth.HttpBearerAuth;
import com.dropbox.sign.model.*;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;

import java.io.File;
import java.util.*;
import java.util.stream.Stream;

public class Requester {
    private final String apiServer;
    private final String authType;
    private final String authKey;
    private JsonNode data;
    private final boolean devMode;
    private JsonNode files;
    private String operationId;
    private JsonNode parameters;
    private final ObjectMapper objectMapper = new ObjectMapper();

    private final static String FILE_UPLOADS_DIR = "./../file_uploads";

    public Requester(String authType, String authKey, String apiServer, String jsonData, boolean devMode) throws Exception {
        this.authType = authType.toLowerCase();
        this.authKey = authKey;
        this.apiServer = apiServer;
        this.devMode = devMode;
        readJsonData(jsonData);
    }

    public void run() throws Exception {
        ObjectWriter ow = objectMapper.writer().withDefaultPrettyPrinter();
        try {
            ApiResponse<?> apiResponse = callFromOperationId();
            Map<String, Object> output = Map.of(
                "body", apiResponse.getData(),
                "status_code", apiResponse.getStatusCode(),
                "headers", apiResponse.getHeaders()
            );
            System.out.println(ow.writeValueAsString(output));
        } catch (ApiException e) {
            Map<String, Object> output = Map.of(
                "body", e.getErrorResponse(),
                "status_code", e.getCode(),
                "headers", e.getResponseHeaders()
            );
            System.out.println(ow.writeValueAsString(output));
        }
    }

    private void readJsonData(String base64Json) throws Exception {
        JsonNode jsonNode = objectMapper.readTree(Base64.getDecoder().decode(base64Json));
        operationId = jsonNode.get("operationId").asText("");
        data = jsonNode.get("data");
        files = jsonNode.get("files");
        parameters = jsonNode.get("parameters");
    }

    private ApiResponse<?> callFromOperationId() throws Exception {
        return Stream.of(
            accountApi(),
            apiAppApi(),
            bulkSendJobApi(),
            embeddedApi(),
            oauthApi(),
            reportApi(),
            signatureRequestApi(),
            teamApi(),
            templateApi(),
            unclaimedDraftApi()
        ).filter(Objects::nonNull).findFirst().orElseThrow(() -> new Exception("Invalid operationId: " + operationId));
    }

    private ApiResponse<?> unclaimedDraftApi() throws Exception {
        UnclaimedDraftApi api = new UnclaimedDraftApi(getApiClient());
        switch (operationId) {
            case "unclaimedDraftCreate":
                UnclaimedDraftCreateRequest createRequest = UnclaimedDraftCreateRequest.init(data.toString());
                createRequest.setFiles(getFiles("files"));
                return api.unclaimedDraftCreateWithHttpInfo(createRequest);
            case "unclaimedDraftCreateEmbedded":
                UnclaimedDraftCreateEmbeddedRequest createEmbeddedRequest =
                    UnclaimedDraftCreateEmbeddedRequest.init(data.toString());
                createEmbeddedRequest.setFiles(getFiles("files"));
                return api.unclaimedDraftCreateEmbeddedWithHttpInfo(createEmbeddedRequest);
            case "unclaimedDraftCreateEmbeddedWithTemplate":
                UnclaimedDraftCreateEmbeddedWithTemplateRequest embeddedWithTemplateRequest =
                    UnclaimedDraftCreateEmbeddedWithTemplateRequest.init(data.toString());
                embeddedWithTemplateRequest.setFiles(getFiles("files"));
                return api.unclaimedDraftCreateEmbeddedWithTemplateWithHttpInfo(embeddedWithTemplateRequest);
            case "unclaimedDraftEditAndResend":
                UnclaimedDraftEditAndResendRequest resendRequest =
                    UnclaimedDraftEditAndResendRequest.init(data.toString());
                return api.unclaimedDraftEditAndResendWithHttpInfo(parameters.get("signature_request_id").asText(), resendRequest);
        }
        return null;
    }

    private ApiResponse<?> templateApi() throws Exception {
        TemplateApi api = new TemplateApi(getApiClient());
        switch (operationId) {
            case "templateAddUser":
                TemplateAddUserRequest addUserRequest = TemplateAddUserRequest.init(data.toString());
                return api.templateAddUserWithHttpInfo(parameters.get("template_id").asText(), addUserRequest);
            case "templateCreateEmbeddedDraft":
                TemplateCreateEmbeddedDraftRequest embeddedDraftRequest
                    = TemplateCreateEmbeddedDraftRequest.init(data.toString());
                return api.templateCreateEmbeddedDraftWithHttpInfo(embeddedDraftRequest);
            case "templateDelete":
                return api.templateDeleteWithHttpInfo(parameters.get("template_id").asText());
            case "templateFilesAsFileUrl":
                return api.templateFilesAsFileUrlWithHttpInfo(
                    parameters.get("template_id").asText()
                );
            case "templateGet":
                return api.templateGetWithHttpInfo(parameters.get("template_id").asText());
            case "templateList":
                return api.templateListWithHttpInfo(
                    parameters.get("account_id").asText(),
                    parameters.get("page").asInt(1),
                    parameters.get("page_size").asInt(20),
                    parameters.get("query").asText()
                );
            case "templateRemoveUser":
                TemplateRemoveUserRequest removeUserRequest = TemplateRemoveUserRequest.init(data.toString());
                return api.templateRemoveUserWithHttpInfo(parameters.get("template_id").asText(), removeUserRequest);
            case "templateUpdateFiles":
                TemplateUpdateFilesRequest updateFilesRequest = TemplateUpdateFilesRequest.init(data.toString());
                updateFilesRequest.setFiles(getFiles("files"));
                return api.templateUpdateFilesWithHttpInfo(parameters.get("template_id").asText(), updateFilesRequest);
        }
        return null;
    }

    private ApiResponse<?> teamApi() throws Exception {
        TeamApi api = new TeamApi(getApiClient());
        switch (operationId) {
            case "teamAddMember":
                TeamAddMemberRequest addMemberRequest = TeamAddMemberRequest.init(data.toString());
                return api.teamAddMemberWithHttpInfo(addMemberRequest, parameters.get("team_id").asText());
            case "teamCreate":
                TeamCreateRequest createRequest = TeamCreateRequest.init(data.toString());
                return api.teamCreateWithHttpInfo(createRequest);
            case "teamDelete":
                return api.teamDeleteWithHttpInfo();
            case "teamGet":
                return api.teamGetWithHttpInfo();
            case "teamRemoveMember":
                TeamRemoveMemberRequest removeMemberRequest = TeamRemoveMemberRequest.init(data.toString());
                return api.teamRemoveMemberWithHttpInfo(removeMemberRequest);
            case "teamUpdate":
                TeamUpdateRequest updateRequest = TeamUpdateRequest.init(data.toString());
                return api.teamUpdateWithHttpInfo(updateRequest);
        }
        return null;
    }

    private ApiResponse<?> signatureRequestApi() throws Exception {
        SignatureRequestApi api = new SignatureRequestApi(getApiClient());
        switch (operationId) {
            case "signatureRequestBulkCreateEmbeddedWithTemplate":
                SignatureRequestBulkCreateEmbeddedWithTemplateRequest templateRequest =
                    SignatureRequestBulkCreateEmbeddedWithTemplateRequest.init(data.toString());
                templateRequest.setSignerFile(getFile("signer_file"));
                return api.signatureRequestBulkCreateEmbeddedWithTemplateWithHttpInfo(templateRequest);
            case "signatureRequestBulkSendWithTemplate":
                SignatureRequestBulkSendWithTemplateRequest withTemplateRequest =
                    SignatureRequestBulkSendWithTemplateRequest.init(data.toString());
                withTemplateRequest.setSignerFile(getFile("signer_file"));
                return api.signatureRequestBulkSendWithTemplateWithHttpInfo(withTemplateRequest);
            case "signatureRequestCancel":
                return api.signatureRequestCancelWithHttpInfo(parameters.get("signature_request_id").asText());
            case "signatureRequestCreateEmbedded":
                SignatureRequestCreateEmbeddedRequest embeddedRequest = SignatureRequestCreateEmbeddedRequest.init(data.toString());
                embeddedRequest.setFiles(getFiles("files"));
                return api.signatureRequestCreateEmbeddedWithHttpInfo(embeddedRequest);
            case "signatureRequestCreateEmbeddedWithTemplate":
                SignatureRequestCreateEmbeddedWithTemplateRequest embeddedWithTemplateRequest =
                    SignatureRequestCreateEmbeddedWithTemplateRequest.init(data.toString());
                embeddedWithTemplateRequest.setFiles(getFiles("files"));
                return api.signatureRequestCreateEmbeddedWithTemplateWithHttpInfo(embeddedWithTemplateRequest);
            case "signatureRequestFilesAsFileUrl":
                return api.signatureRequestFilesAsFileUrlWithHttpInfo(
                    parameters.get("signature_request_id").asText()
                );
            case "signatureRequestGet":
                return api.signatureRequestGetWithHttpInfo(parameters.get("signature_request_id").asText());
            case "signatureRequestList":
                return api.signatureRequestListWithHttpInfo(
                    parameters.get("account_id").asText(),
                    parameters.get("page").asInt(1),
                    parameters.get("page_size").asInt(20),
                    parameters.get("query").asText()
                );
            case "signatureRequestReleaseHold":
                return api.signatureRequestReleaseHoldWithHttpInfo(parameters.get("signature_request_id").asText());
            case "signatureRequestRemind":
                SignatureRequestRemindRequest remindRequest = SignatureRequestRemindRequest.init(data.toString());
                return api.signatureRequestRemindWithHttpInfo(parameters.get("signature_request_id").asText(), remindRequest);
            case "signatureRequestRemove":
                return api.signatureRequestRemoveWithHttpInfo(parameters.get("signature_request_id").asText());
            case "signatureRequestSend":
                SignatureRequestSendRequest sendRequest = SignatureRequestSendRequest.init(data.toString());
                sendRequest.setFiles(getFiles("files"));
                return api.signatureRequestSendWithHttpInfo(sendRequest);
            case "signatureRequestSendWithTemplate":
                SignatureRequestSendWithTemplateRequest sendWithTemplateRequest =
                    SignatureRequestSendWithTemplateRequest.init(data.toString());
                sendWithTemplateRequest.setFiles(getFiles("files"));
                return api.signatureRequestSendWithTemplateWithHttpInfo(sendWithTemplateRequest);
            case "signatureRequestUpdate":
                SignatureRequestUpdateRequest updateRequest = SignatureRequestUpdateRequest.init(data.toString());
                return api.signatureRequestUpdateWithHttpInfo(parameters.get("signature_request_id").asText(), updateRequest);

        }
        return null;
    }

    private ApiResponse<?> reportApi() throws Exception {
        ReportApi api = new ReportApi(getApiClient());
        if (operationId.equals("reportCreate")) {
            ReportCreateRequest createRequest = ReportCreateRequest.init(data.toString());
            return api.reportCreateWithHttpInfo(createRequest);
        }
        return null;
    }

    private ApiResponse<?> oauthApi() throws Exception {
        OAuthApi api = new OAuthApi(getApiClient());
        switch (operationId) {
            case "oauthTokenGenerate":
                OAuthTokenGenerateRequest generateRequest = OAuthTokenGenerateRequest.init(data.toString());
                return api.oauthTokenGenerateWithHttpInfo(generateRequest);
            case "oauthTokenRefresh":
                OAuthTokenRefreshRequest refreshRequest = OAuthTokenRefreshRequest.init(data.toString());
                return api.oauthTokenRefreshWithHttpInfo(refreshRequest);
        }
        return null;
    }

    private ApiResponse<?> embeddedApi() throws Exception {
        EmbeddedApi api = new EmbeddedApi(getApiClient());
        switch (operationId) {
            case "embeddedEditUrl":
                EmbeddedEditUrlRequest editUrlRequest = EmbeddedEditUrlRequest.init(data.toString());
                return api.embeddedEditUrlWithHttpInfo(parameters.get("template_id").asText(), editUrlRequest);
            case "embeddedSignUrl":
                return api.embeddedSignUrlWithHttpInfo(parameters.get("embeddedSignUrl").asText());
        }
        return null;
    }

    private ApiResponse<?> bulkSendJobApi() throws Exception {
        BulkSendJobApi api = new BulkSendJobApi(getApiClient());
        switch (operationId) {
            case "bulkSendJobGet":
                return api.bulkSendJobGetWithHttpInfo(parameters.get("bulk_send_job_id").asText());
            case "bulkSendJobList":
                int page = parameters.get("page").asInt(1);
                int pageSize = parameters.get("page_size").asInt(20);
                return api.bulkSendJobListWithHttpInfo(page, pageSize);
        }
        return null;
    }

    private ApiResponse<?> apiAppApi() throws Exception {
        ApiAppApi api = new ApiAppApi(getApiClient());
        switch (operationId) {
            case "apiAppCreate":
                ApiAppCreateRequest createRequest = ApiAppCreateRequest.init(data.toString());
                createRequest.customLogoFile(getFile("custom_logo_file"));
                return api.apiAppCreateWithHttpInfo(createRequest);
            case "apiAppDelete":
                return api.apiAppDeleteWithHttpInfo(parameters.get("client_id").asText());
            case "apiAppGet":
                return api.apiAppGetWithHttpInfo(parameters.get("client_id").asText());
            case "apiAppList":
                int page = parameters.get("page").asInt(1);
                int pageSize = parameters.get("page_size").asInt(20);
                return api.apiAppListWithHttpInfo(page, pageSize);
            case "apiAppUpdate":
                ApiAppUpdateRequest updateRequest = ApiAppUpdateRequest.init(data.toString());
                updateRequest.customLogoFile(getFile("custom_logo_file"));
                return api.apiAppUpdateWithHttpInfo(parameters.get("client_id").asText(), updateRequest);
        }
        return null;
    }

    private ApiResponse<?> accountApi() throws Exception {
        AccountApi api = new AccountApi(getApiClient());
        switch (operationId) {
            case "accountCreate":
                AccountCreateRequest request = AccountCreateRequest.init(data.toString());
                return api.accountCreateWithHttpInfo(request);
            case "accountGet":
                return api.accountGetWithHttpInfo(
                    Optional.ofNullable(parameters.get("account_id")).map(JsonNode::asText).orElse(null),
                    Optional.ofNullable(parameters.get("email_address")).map(JsonNode::asText).orElse(null)
                );
            case "accountUpdate":
                AccountUpdateRequest updateRequest = AccountUpdateRequest.init(data.toString());
                return api.accountUpdateWithHttpInfo(updateRequest);
            case "accountVerify":
                AccountVerifyRequest verifyRequest = AccountVerifyRequest.init(data.toString());
                return api.accountVerifyWithHttpInfo(verifyRequest);
        }
        return null;
    }

    private ApiClient getApiClient() throws Exception {
        ApiClient defaultClient = Configuration.getDefaultApiClient();
        if (devMode) {
            defaultClient.addDefaultHeader("Cookie", "XDEBUG_SESSION=xdebug");
        }
        if (apiServer != null && !apiServer.isBlank()) {
            defaultClient.setBasePath("https://" + apiServer + "/v3");
        }
        if ("apikey".equals(authType)) {
            HttpBasicAuth api_key = (HttpBasicAuth) defaultClient
                .getAuthentication("api_key");
            api_key.setUsername(authKey);
        } else if ("oauth".equals(authType)) {
            HttpBearerAuth oauth2 = (HttpBearerAuth) defaultClient
                .getAuthentication("oauth2");

            oauth2.setBearerToken(authKey);
        } else {
            throw new Exception(
                "Invalid auth type. Must be \"apikey\" or \"oauth\""
            );
        }
        return defaultClient;
    }

    private List<File> getFiles(String name) {
        List<File> files = new ArrayList<>();
        this.files.get(name)
            .iterator()
            .forEachRemaining(jsonNode -> files.add(new File(FILE_UPLOADS_DIR + "/" + jsonNode.asText())));
        return files;
    }

    private File getFile(String name) {
        if (!files.get(name).asText().isBlank()) {
            return new File(FILE_UPLOADS_DIR + "/" + files.get(name).asText());
        }
        return null;
    }

    public static void main(String[] args) throws Exception {
        Requester r = new Requester(
            System.getenv("AUTH_TYPE"),
            System.getenv("AUTH_KEY"),
            System.getenv("API_SERVER"),
            System.getenv("JSON_DATA"),
            Boolean.parseBoolean(System.getenv("DEV_MODE"))
        );
        r.run();
    }
}
